### PR TITLE
Fix various RLE & buffer related bugs

### PR DESCRIPTION
**Pull request by Katharsas**
- change buffering implementation to work with RLE compression. Remaining buffer will be tested on each pixel instead of row because its not possible to predict buffer usage thanks to RLE)
- fixed wrong runLength decoding for RLE compressed files (regression from https://github.com/tmyroadctfig/com.realityinteractive.imageio.tga/pull/5/commits/71d690317b26d1504a1e7d9c558dc259bf04e73d)
- fixed remaining bytes getting lost during buffer refill (this is relevant when buffersize dos not allign with input pixel size, usually the case for RLE compression)
- ~(change from bits per pixel to bytes per pixel for decoding switch)~

Buffer size is now fixed and not a multiple of row size. It can be any size but should be multiple of 3 and 4 so that it is alligned with input pixel sizes for non-RLE-compressed files (makes buffer refill faster).

I guess unit tests on various differently encoded TGAs would really help to prevent my PRs breaking stuff^^. I might look in to that.

### CODE REVIEW

**Comment by tmyroadctfig** [comment]
It is possible to add in a simple unit test to sanity check this change?

File: src/main/java/com/realityinteractive/imageio/tga/TGAImageReader.java [hunk 1]
```diff
@@ -440,16 +429,20 @@ public BufferedImage read(final int imageIndex, final ImageReadParam param)
                         readPixel = isRaw;
                     } else /* non-positive run length */
                     {
+                        // make sure we have enough input in buffer, return if end of file is reached
+                        if (checkFillBuffer(inputStream, inputBuffer, bytesPerPixel + 1))
+                            return image;
+                        
                         // read the repetition count field 
-                        runLength = inputBuffer.get();
+                        runLength = inputBuffer.get() & 0xFF; // unsigned
```
Comment (Katharsas) [inline src/main/java/com/realityinteractive/imageio/tga/TGAImageReader.java]: it should be much clearer now what the problem was.

File: src/main/java/com/realityinteractive/imageio/tga/TGAImageReader.java [hunk 2]
```diff
@@ -440,16 +429,20 @@ public BufferedImage read(final int imageIndex, final ImageReadParam param)
...
 
                         // determine which packet type:  raw or runlength
                         isRaw = ( (runLength & 0x80) == 0); // bit 7 == 0 -> raw; bit 7 == 1 -> runlength
 
-                        // if a run length packet then shift to get the number
+                        // if a run length packet then set bit 7 to 0 to get length
                         if(!isRaw)
                             runLength -= 0x80;
-                        /* else -- is raw so there's no need to shift */ 
+                        /* else -- is raw so there's no need remove bit 7 */
```
Comment (Katharsas) [inline src/main/java/com/realityinteractive/imageio/tga/TGAImageReader.java]: Removed binary literals to keep things the same.

**Comment by Katharsas** [comment]
I'm investigating unit tests. I do not want to make the images i used myself for testing public. Do you know a free program that can save tga in various programs?

Btw i reverted `(change from bits per pixel to bytes per pixel for decoding switch)` just now and also properly redid the RLE encoding fix. It should be much clearer now why/how it works (see comment)

**Comment by Katharsas** [comment]
Ok, i should be able to create unit tests for 16/24/32 bit with/without RLE but i don't have an external program that can create color mapped or 8bit tgas i think, so ill leave that out. Will take some weeks though

**Comment by Katharsas** [comment]
Ok, the new "integration" test revealed another regression from my last PR and a general bug that i think has always existed with how 5 bit colors where scaled to 6 bit (old code did not make sure that full black is converted to full black and full white to full white).

I currently have no way to test 16 bit with 1 bit alpha and 8 bit, but if i get my hands on a good encoder in a feature it will be trivial to add that to the integration test.

It may also be good to upgrade gradle wrapper version because otherwise "gradle test" does not show line number where test fails by default, but that can be a different PR i guess.

**Comment by tmyroadctfig** [comment]
Thanks for cleaning up the existing test, and adding in the new test. 👍

**Pull request merged**